akka-sample-kafka-to-sharding-scala/processor/src/main/scala/sample/sharding/kafka/UserEventsKafkaProcessor.scala

```diff
@@ -29,35 +29,42 @@ object UserEventsKafkaProcessor {
         implicit val ec: ExecutionContextExecutor = ctx.executionContext
         implicit val scheduler: Scheduler = classic.scheduler
 
+        // ActorContext can't be accessed inside Futures
+        // so we capture the scheduler and log before using them
+        val typedScheduler = ctx.system.scheduler
+        val logger = ctx.log
+
         val rebalanceListener = KafkaClusterSharding(classic).rebalanceListener(processorSettings.entityTypeKey)
 
         val subscription = Subscriptions
           .topics(processorSettings.topics: _*)
           .withRebalanceListener(rebalanceListener.toClassic)
 
-        val stream: Future[Done] = Consumer.sourceWithOffsetContext(processorSettings.kafkaConsumerSettings(), subscription)
-          // MapAsync and Retries can be replaced by reliable delivery
-          .mapAsync(20) { record =>
-            ctx.log.info(s"user id consumed kafka partition ${record.key()}->${record.partition()}")
-            retry(() =>
-              shardRegion.ask[Done](replyTo => {
-                val purchaseProto = UserPurchaseProto.parseFrom(record.value())
-                UserEvents.UserPurchase(
-                  purchaseProto.userId,
-                  purchaseProto.product,
-                  purchaseProto.quantity,
-                  purchaseProto.price,
-                  replyTo)
-              })(processorSettings.askTimeout, ctx.system.scheduler),
-              attempts = 5,
-              delay = 1.second
-            )
-          }
-          .runWith(Committer.sinkWithOffsetContext(CommitterSettings(classic)))
+        val stream: Future[Done] =
```

Comment:
Perhaps a good time to move this out to a static method, that will make it explicit what dependencies the future callbacks, it's what we have said to prefer internally in team and I think also in styleguide, to avoid closing over mutable actor state in future callbacks in general.

An additional thought, if you'd add mdc to this actor and still use the logger in other threads I'm not entirely sure what will happen, some log events may get mdc values from the actor and some not, so perhaps it should be a separate logger?

Reply:
I agree. Future and streams stuff should be defined outside the actor.

I also agree that it should use a separate logger, or maybe better the Streams `log` operator.